### ISSUE TITLE
LDValue prevents easy encoding/decoding of json objects into local models

### ISSUE THREAD

**Issue by mcmurrym** [launchdarkly/ios-client#274]
When requesting a json variation, the dictionary now represents all values as an `LDValue`. This prevents turning the value into JSON Data and then using `JSONDecoder().decode(MyOwnType.self, jsonVariationData)` in previous versions the type retrieved via generics, this is much preferred and allow easy encoding/decoding into our own local models. Can we please get a way to get a more raw value back that matches exactly what was set?

**Comment by mcmurrym** [comment]
https://github.com/launchdarkly/ios-client-sdk/issues/275

**Comment by mkambale** [comment]
@mcmurrym As a swift developer, I complete echo it.
Its painful to work work with LDValues when we have json data.

**Comment by louis-launchdarkly** [comment]
Hello @mcmurrym and @mkambale, thank you for reaching out and explaining your use cases. Let me discuss this with the engineers on the team.

_Filed Internally as 160632._

**Comment by mcmurrym** [comment]
Any update on this @louis-launchdarkly? We are still on an sdk version prior to the LDValue change.

**Comment by keelerm84** [comment]
I don't have an update on the timeline for the ticket, but I do have a question about this report so I can better understand your needs.

> When requesting a json variation, the dictionary now represents all values as an LDValue. This prevents turning the value into JSON Data and then using JSONDecoder().decode(MyOwnType.self, jsonVariationData)

The `LDValue` enum is `Codable`, which means you can JSON encode and decode it directly. If your goal is to get the JSON variation result back as a JSON string, would something like this be sufficient?

```swift
let detail = client.jsonVariationDetail(forKey: "json-variation", defaultValue: .null)
let json_string = try JSONEncoder().encode(detail.value)
print(String(data: json_string, encoding: .utf8))
```

**Comment by mcmurrym** [comment]
Hmmm, yes, that does work. I've updated to 8.0 and am using this code. @mkambale Maybe this works for you as well?
Thanks for the tip. Obviously the roundtrip isn't ideal, but acceptable for now.

```
func object<T: Decodable>(from key: String, defaultValue: T) -> T {
    guard let client = client else { return defaultValue }
    let json = client.jsonVariation(forKey: key, defaultValue: .null)
    if json == .null {
        return defaultValue
    }
    
    do {
        let encoded = try JSONEncoder().encode(json)
        return try JSONDecoder().decode(T.self, from: encoded)
    } catch {
        return defaultValue
    }
}
```

**Comment by JoeSzymanskiFAN** [comment]
I have hit a case where this would still be preferable as a feature. In our case, we are using Launch Darkly in some Objective C code, so we don't have access to `Codable` at all. In our case, having access to the raw JSON string would be the simplest and most performant solution, even if it was just allowing a JSON flag to be access as a string variation somehow.

**Comment by JoeSzymanskiFAN** [comment]
As a separate option, the shared Java SDK has a method to convert an LDValue into a JSON string, which would be nice to have on the iOS side also: https://github.com/launchdarkly/java-sdk-common/blob/main/src/main/java/com/launchdarkly/sdk/LDValue.java#L384-L394

**Comment by brianmichel** [comment]
Hey folks, any update on this? I just ran into this myself as we've built a system that will automatically convert object variations into our own Codable structs and having all of the values be `LDValue`s ends up breaking this conversion entirely.

It would be ideal to have some function that returns array variations as `[Any]` and object variations as `[String: Any]` so that the consumers of these APIs can build whatever they would like without worrying about have to first confirm from LD specific values.

**Comment by keelerm84** [comment]
We just released v9.4.0 which adds two new methods `variation<T>` and `variationDetail<T>`! Both of these methods accept a type that must be Decodable and implement our LDValueConvertible protocol (required for now to deal with some internal plumbing).
